### PR TITLE
[generator] Fix bug 43579 - Generator emits invalid code when using the same method name (overload) in @delegates using events and C# delegates

### PR DESCRIPTION
**Pull request by dalexsoto**
https://bugzilla.xamarin.com/show_bug.cgi?id=43579

Bug Description:
Generator will emit invalid code when using the same name (overload)
in a method inside a @delegate (protocol) that is decorated
either with EventArgs or DelegateName

-----------------------------------------------------------------

This commit introduces a new attribute named `DelegateApiNameAttribute`
which mimics the `EventNameAttribute` but for delegate properties in
host classes

It is used to specify the delegate property name that will be created when
the generator creates the delegate property on the host
class that holds events and delegates.

This is really useful when you have two overload methods that makes
sense to keep them named as is but you want to expose them in the host class
with a better given name.

example:
```
 interface SomeDelegate {
     [Export ("foo"), DelegateApiName ("Confirmation"), DelegateName ("Func<bool>"), DefaultValue (false)]
     bool Confirm (Some source);
 }
```

 Generates propety in the host class:
	`Func<bool> Confirmation { get; set; }`

This also introduces two new BIXXXX errors:
- BI1043 Repeated overload {mi.Name} and no [DelegateApiNameAttribute]
provided to generate property name on host class.
- BI1044 Repeated name '{apiName.Name}' provided in [DelegateApiNameAttribute].

Which provides an error instead of generating invalid C# code.

Generator test included :D

Build Diff: https://gist.github.com/dalexsoto/647e3d4a97e07ef612475b56054b8e4e

### CODE REVIEW

File: src/error.cs [hunk 1]
```diff
@@ -57,6 +57,8 @@
 //		BI1040 The selector {0} on type {1} is found multiple times with different argument out states on argument {2}.
 //		BI1041 The selector {0} on type {1} is found multiple times with different argument types on argument {2} - {3} : {4}.
 //		BI1042 Missing '[Field (LibraryName=value)]' for {field_pi.Name} (e.g."__Internal")
+//		BI1043 Repeated overload {mi.Name} and no [DelegateApiNameAttribute] provided to generate property name on host class.
+//		BI1044 Repeated name '{apiName.Name}' provided in [DelegateApiNameAttribute].
```
Comment (spouliot) [inline src/error.cs]: we don't yet have a a doc file for `BIxxxx` (so it's fine to have the errors just here) but...

File: src/generator.cs [hunk 2]
```diff
@@ -707,6 +707,34 @@ public DelegateNameAttribute (string s)
...
+// interface SomeDelegate {
+//     [Export ("foo"), DelegateApiName ("Confirmation"), DelegateName ("Func<bool>"), DefaultValue (false)]
+//     bool Confirm (Some source);
+// }
+//
+// Generates propety in the host class:
+//	Func<bool> Confirmation { get; set; }
+//
+//
+public class DelegateApiNameAttribute : Attribute {
```
Comment (spouliot) [inline src/generator.cs]: The commit needs to update `docs/website/binding_types_reference_guide.md` too

File: src/generator.cs [hunk 3]
```diff
@@ -7301,6 +7339,25 @@ string GetEventName (MethodInfo mi)
 		return ea.EvtName;
 	}
 
+	List<string> repeatedDelegateApiNames = new List<string> ();
```
Comment (spouliot) [inline src/generator.cs]: maybe an `HashList` would be faster ?

File: tests/generator/Makefile [hunk 4]
```diff
@@ -13,7 +13,7 @@ include $(TOP)/Make.config
 
 IOS_CURRENT_DIR=$(IOS_DESTDIR)/Library/Frameworks/Xamarin.iOS.framework/Versions/Current
 IOS_GENERATOR = $(IOS_CURRENT_DIR)/bin/btouch /baselib:$(IOS_CURRENT_DIR)/lib/mono/2.1/monotouch.dll /unsafe /compiler:$(IOS_CURRENT_DIR)/bin/smcs
-IOS_TESTS = bug15283 bug15307 bug15799 bug16036 sof20696157 bug23041 bug27430 bug27428 bug34042 btouch-with-hyphen-in-name property-redefination-ios arrayfromhandlebug bug36457 bug39614 bug40282 bug17232 bug24078-ignore-methods-events strong-dict-support-templated-dicts
+IOS_TESTS = bug15283 bug15307 bug15799 bug16036 sof20696157 bug23041 bug27430 bug27428 bug34042 btouch-with-hyphen-in-name property-redefination-ios arrayfromhandlebug bug36457 bug39614 bug40282 bug17232 bug24078-ignore-methods-events strong-dict-support-templated-dicts bug43579
```
Comment (spouliot) [inline tests/generator/Makefile]: ❤️

File: src/generator.cs (hunk 2 above)
Comment (dalexsoto) [inline src/generator.cs]: Done in https://github.com/xamarin/xamarin-macios/pull/817/commits/e5d584e8bf0c01bbcd0795b2e1a431cc6f6b1c1e

**Pull request merged**